aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,394 @@
+    ///
+    /// `alg` is the `KemAlgorithm` to be associated with the generated `KemPrivateKey`
+    ///
+    /// `bytes` is a slice of raw bytes representing a `KemPrivateKey`
+    ///
+    /// # Errors
+    /// `error::KeyRejected` when operation fails during key creation.
+    ///
+    pub fn new(alg: KemAlgorithm, bytes: &[u8]) -> Result<Self, KeyRejected> {
+        unsafe {
+            let privkey_ctx = LcPtr::new(EVP_PKEY_kem_new_raw_secret_key(
+                alg.nid(),
+                bytes.as_ptr(),
+                bytes.len(),
+            ))?;
+            Ok(KemPrivateKey {
+                algorithm: alg,
+                context: privkey_ctx,
+                priv_key: bytes.to_owned().into(),
+            })
+        }
+    }
+}
+
+impl AsRef<[u8]> for KemPrivateKey {
+    fn as_ref(&self) -> &[u8] {
+        &self.priv_key
+    }
+}
+
+/// A serializable public key usable with KEMS. This can be constructed
+/// from a `KemPrivateKey` or constructed from raw bytes.
+#[derive(Debug)]
+pub struct KemPublicKey {
+    algorithm: KemAlgorithm,
+    context: LcPtr<*mut EVP_PKEY>,
+    pub_key: Box<[u8]>,
+}
+
+impl KemPublicKey {
+    /// Return the algorithm associated with the given KEM public key.
+    #[must_use]
+    pub fn algorithm(&self) -> &KemAlgorithm {
+        &self.algorithm
+    }
+
+    /// Performs the encapsulate operation using the current KEM public key, generating a ciphertext
+    /// and associated shared secret.
+    ///
+    /// After the decapsulation is finished, `decapsulate` calls `kdf` with the ciphertext and raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn encapsulate<F, R>(&self, kdf: F) -> Result<R, Unspecified>
+    where
+        F: FnOnce(&[u8], &[u8]) -> Result<R, Unspecified>,
+    {
+        unsafe {
+            let ctx = LcPtr::new(EVP_PKEY_CTX_new(*self.context, null_mut()))?;
+            let mut ciphertext_len: usize = 0;
+            let mut shared_secret_len: usize = 0;
+
+            if 1 != EVP_PKEY_encapsulate(
+                *ctx,
+                null_mut(),
+                &mut ciphertext_len,
+                null_mut(),
+                &mut shared_secret_len,
+            ) {
+                ciphertext_len.zeroize();
+                shared_secret_len.zeroize();
+                return Err(Unspecified);
+            }
+
+            let mut ciphertext: Vec<u8> = vec![0u8; ciphertext_len];
+            let mut shared_secret: Vec<u8> = vec![0u8; shared_secret_len];
+
+            if EVP_PKEY_encapsulate(
+                *ctx,
+                ciphertext.as_mut_ptr(),
+                &mut ciphertext_len,
+                shared_secret.as_mut_ptr(),
+                &mut shared_secret_len,
+            ) != 1
+            {
+                ciphertext.zeroize();
+                shared_secret.zeroize();
```

Comment:
* Not quite sure if these need to be zerioized either. Same for above.
* Zeroization is great in that we don't leak any potential private data. But I think it should be more valuable when we already have secret values capable of being leaked. If we do want to implement Zeroization, something like what we're doing for HKDF below would be more valuable. This helps ensure that any available secret data we have is zeroized before it's being freed/dropped. 
* https://github.com/aws/aws-lc-rs/blob/832e839e9013d96775df726df32f566e4558b3bd/aws-lc-rs/src/hkdf.rs#L104-L108